linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1268 @@
+            "Number of requests that were served from cache",
+        )
+    });
+}
+
+/// Manages a pool of validator nodes with intelligent load balancing and performance tracking.
+///
+/// The `ValidatorManager` maintains performance metrics for each validator node using
+/// Exponential Moving Averages (EMA) and uses these metrics to make intelligent routing
+/// decisions. It prevents node overload through request capacity limits and automatically
+/// retries failed requests on alternative nodes.
+///
+/// # Examples
+///
+/// ```ignore
+/// // Create with default configuration (balanced scoring)
+/// let manager = ValidatorManager::new(validator_nodes);
+///
+/// // Create with custom configuration prioritizing low latency
+/// let latency_weights = ScoringWeights {
+///     latency: 0.6,
+///     success: 0.3,
+///     load: 0.1,
+/// };
+/// let manager = ValidatorManager::with_config(
+///     validator_nodes,
+///     15,                      // max 15 concurrent requests per node
+///     latency_weights,         // custom scoring weights
+///     0.2,                     // higher alpha for faster adaptation
+///     3000.0,                  // max expected latency (3 seconds)
+///     Duration::from_secs(60), // 60 second cache TTL
+///     200,                     // cache up to 200 entries
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct ValidatorManager<Env: Environment> {
+    /// Thread-safe map of validator nodes indexed by their public keys.
+    /// Each node is wrapped with EMA-based performance tracking information.
+    nodes: Arc<tokio::sync::RwLock<BTreeMap<ValidatorPublicKey, NodeInfo<Env>>>>,
+    /// Maximum number of concurrent requests allowed per node.
+    /// Prevents overwhelming individual validators with too many parallel requests.
+    max_requests_per_node: usize,
+    /// Default scoring weights applied to new nodes.
+    default_weights: ScoringWeights,
+    /// Default EMA smoothing factor for new nodes.
+    default_alpha: f64,
+    /// Default maximum expected latency in milliseconds for score normalization.
+    default_max_expected_latency_ms: f64,
+    /// Tracks in-flight requests to deduplicate concurrent requests for the same data.
+    in_flight_tracker: InFlightTracker<RemoteNode<Env::ValidatorNode>>,
+    /// Cache of recently completed requests with their results and timestamps.
+    cache: RequestsCache<RequestKey, RequestResult>,
+}
+
+impl<Env: Environment> ValidatorManager<Env> {
+    /// Creates a new `ValidatorManager` with default configuration.
+    pub fn new(nodes: impl IntoIterator<Item = RemoteNode<Env::ValidatorNode>>) -> Self {
+        Self::with_config(
+            nodes,
+            MAX_IN_FLIGHT_REQUESTS,
+            ScoringWeights::default(),
+            0.1,
+            MAX_ACCEPTED_LATENCY_MS,
+            Duration::from_secs(CACHE_TTL_SEC), // 60 second cache TTL
+            CACHE_MAX_SIZE,                     // cache up to 100 entries
```

Comment:
Those comments will become outdated if we change the default values. If I did, I wouldn't think of looking into this function.